### ISSUE TITLE
support for batch inserts

### ISSUE THREAD

**Issue by yogthos** [krisajenkins/yesql#51]
As far as I can tell something like the following is not currently possible with Yesql syntax:

```sql
INSERT INTO products (name, price) VALUES
    ('Cheese', 9.99),
    ('Bread', 1.99),
    ('Milk', 2.99);
```
Perhaps a hint could be used to indicate that the function accepts a collection of arguments where each item represents a row to be inserted:

```sql
--name: create-products!
--batch: true
INSERT INTO products
(product_no, name, price) VALUES
VALUES (:first_name, :last_name, :email, :pass)
```
```clojure
(create-products! [["Cheese" 9.99] 
                             ["Bread" 1.99]
                             ["Milk" 2.99]])
```

Alternatively, the function could check if it's passed a collection and handle that intelligently. Downside being that it's a bit more magical.

**Comment by ryanbaldwin** [comment]
I'm assumping the :product_no in the above example is a typo. Again. C'mon Dmitri! ;)

Anyways - I think an implicit handling of lists is better than annotating statements with _batch_ because if _batch_ is strict (ie, it only allows a list) then you're forced to either 
a) write the same exact sql statement twice and annotate it differently (one with batch, one without), or 
b) do something like the following (create-products! [["Cheese" 9.99]]), which is weird.

Option b uglifies, and option a is duplicate code.

With YeSQL 0.5.* I think an implicit handling of vector vs map is rad (whereas in 0.4.0 you'd have to do the painful vector of vectors checking). 

In summary, I prefer this (0.5.0 syntax):
```clojure
(create-products! {:name "Cheese" :price 9.99})
(create-products! [{:name "Cheese" :price 9.99} {:name "Bread" :price 1.99}...]
```

**Comment by yogthos** [comment]
fixed the typos in the example :)

I think I'm also leaning towards the implicit syntax, especially in context of `0.5.0` using maps for records. It might actually be the cleanest to simply use varargs:

```clojure
(create-products! {:name "Cheese" :price 9.99} {:name "Bread" :price 1.99} ...)
```

**Comment by ryanbaldwin** [comment]
I like the varargs approach too. 

- ryan.

**Comment by yogthos** [comment]
Also worth noting that `clojure.java.jdbc` already supports this behavior:

```clojure
(clojure.java.jdbc/insert! db :users
   {:id "foo1" :pass "bar"}
   {:id "foo2" :pass "bar"}
   {:id "foo3" :pass "bar"})
```

**Comment by yogthos** [comment]
Another note is that the connection is passed in as the last parameter with the 0.5.x syntax, it might be better to pass it in as the first to avoid ambiguity for batched parameters.

**Comment by ryanbaldwin** [comment]
It used to be the first parameter, and it was terribly annoying as you had to pass it around all over the place. In 0.5.x you pass the connection in as part of the def when you load the SQL, and after that it's optional. That optional, however, is a map, so it wouldn't be much to extend that map if required, but that seems like a bit of a hack. 



- ryan.

**Comment by yogthos** [comment]
I don't think the two things are at odds with each other. It can still be the first parameter, and the `defqueries` macro can implicitly insert it when provided.

**Comment by dhruvbhatia** [comment]
+1 batch inserts/updates would be very handy. The varargs approach looks cleanest.

**Comment by yogthos** [comment]
@krisajenkins any news on whether there are plans to support this, it's been a year :)

**Comment by prateekbhatt** [comment]
Ya, supporting batch inserts/updates would be very useful. 

@yogthos What workarounds do we have now to do batch inserts ? I am using Luminus with yesql. Thanks

**Comment by yogthos** [comment]
@prateekbhatt I end up having to drop back to using `clojure.java.jdbc` directly and it feels a bit silly since that's what yesql is wrapping

**Comment by michaelbarton** [comment]
I think this feature would certainly be useful. At the moment I have to do batch inserts in my clojure code when I would rather push this to the SQL layer of my code so I can combine this with features like UPSERT.

**Comment by iku000888** [comment]
+1

This seriously needs to be implemented!
The intelligent implicit syntax is better.

**Comment by jgeraerts** [comment]
+1
